File: .spec/scripts/python/create_new_feature.py
#!/usr/bin/env python3
"""
create_new_feature.py — Bootstrap a new feature: derives branch/folder names,
creates the git branch and specs/YYYYMMDD-<name>/ directory, and keeps
session.json in sync (analogous to create-new-feature.ps1).

Naming conventions
  branch  : <name>  →  <name>-YYYYMMDD  →  <name>-YYYYMMDD-2 …
  folder  : YYYYMMDD-<name>  →  YYYYMMDD-<name>-2 …  (date-prefix for dir sorting)

Usage:
    python create_new_feature.py --name "oauth2-login" --description "Implements OAuth2 login"
                                  [--agent-name "spec.specify"] [--json] [--dry-run]
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from datetime import datetime
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from common import get_repo_root, test_has_git, resolve_template  # noqa: E402

SCRIPT_DIR = Path(__file__).parent


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _run_manage(args: list[str], capture: bool = False):
    """Run manage_session.py and return (returncode, stdout)."""
    cmd = [sys.executable, str(SCRIPT_DIR / "manage_session.py")] + args
    if capture:
        result = subprocess.run(cmd, capture_output=True, text=True)
        return result.returncode, result.stdout.strip()
    rc = subprocess.run(cmd).returncode
    return rc, ""


def _log(msg: str) -> None:
    print(f"[feature] {msg}")


def _git_run(args: list[str], repo_root: str, capture: bool = False):
    cmd = ["git", "-C", repo_root] + args
    if capture:
        result = subprocess.run(cmd, capture_output=True, text=True)
        return result.returncode, result.stdout.strip()
    rc = subprocess.run(cmd, capture_output=True).returncode
    return rc, ""


# ---------------------------------------------------------------------------
# Unique name generators
# ---------------------------------------------------------------------------

def get_unique_branch_name(base: str, repo_root: str, has_git: bool) -> str:
    taken: set[str] = set()
    if has_git:
        rc, out = _git_run(["branch", "--list", f"{base}*"], repo_root, capture=True)
        if rc == 0:
            for line in out.splitlines():
                taken.add(line.strip().lstrip("*").strip())
        rc, out = _git_run(["branch", "-r", "--list", f"*/{base}*"], repo_root, capture=True)
        if rc == 0:
            for line in out.splitlines():
                taken.add(line.strip().split("/")[-1])

    if base not in taken:
        return base
    ds = datetime.now().strftime("%Y%m%d")
    dc = f"{base}-{ds}"
    if dc not in taken:
        return dc
    n = 2
    while True:
        c = f"{base}-{ds}-{n}"
        if c not in taken:
            return c
        n += 1


def get_unique_folder_name(base: str, specs_dir: Path) -> str:
    ds     = datetime.now().strftime("%Y%m%d")
    prefix = f"{ds}-{base}"
    taken: set[str] = set()
    if specs_dir.is_dir():
        for d in specs_dir.iterdir():
            if d.is_dir() and d.name.startswith(prefix):
                taken.add(d.name)
    if prefix not in taken:
        return prefix
    n = 2
    while True:
        c = f"{prefix}-{n}"
        if c not in taken:
            return c
        n += 1


# ---------------------------------------------------------------------------
# Output helper
# ---------------------------------------------------------------------------

def write_result(branch: str, feat_dir: str, use_json: bool) -> None:
    if use_json:
        print(json.dumps({
            "BRANCH_NAME": branch,
            "FEATURE_DIR": feat_dir,
            "SPEC_FILE":   f"{feat_dir}/spec.md",
        }, ensure_ascii=False))
    else:
        print(f"BRANCH_NAME: {branch}")
        print(f"FEATURE_DIR: {feat_dir}")
        print(f"SPEC_FILE:   {feat_dir}/spec.md")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Bootstrap a new feature branch and spec directory."
    )
    parser.add_argument("--name",        default="", help="Feature name slug (e.g. oauth2-login-google)")
    parser.add_argument("--description", default="", help="One-line feature description")
    parser.add_argument("--agent-name",  default="", help="Calling agent to record in session")
    parser.add_argument("--json",        action="store_true", help="Output JSON")
    parser.add_argument("--dry-run",     action="store_true", help="Compute names without creating anything")
    args = parser.parse_args()

    repo_root  = get_repo_root()
    has_git    = test_has_git(repo_root)
    specs_dir  = Path(repo_root) / "specs"

    if not args.dry_run:
        specs_dir.mkdir(parents=True, exist_ok=True)

    # ── Session bootstrap ────────────────────────────────────────────────────
    if not args.dry_run:
        init_args = ["--action", "init"]
        if args.name:        init_args += ["--name",        args.name]
        if args.description: init_args += ["--description", args.description]
        rc, _ = _run_manage(init_args)
        if rc != 0:
            sys.exit(rc)

    # Read current session values
    base_name     = ""
    sess_branch   = ""
    sess_feat_dir = ""

    rc, raw = _run_manage(["--action", "get-multi", "--fields", "name,branch_name,feature_dir"], capture=True)
    if rc == 0 and raw:
        try:
            sess_data     = json.loads(raw)
            base_name     = sess_data.get("name", "") or ""
            sess_branch   = sess_data.get("branch_name", "") or ""
            sess_feat_dir = sess_data.get("feature_dir", "") or ""
        except Exception:
            pass

    if not base_name:
        base_name = args.name
    if not base_name:
        msg = (
            "ERROR: No active session and --name not provided. Pass --name <slug> for dry-run."
            if args.dry_run else
            "ERROR: Feature name is not set. Pass --name <slug> to provide one."
        )
        _log(msg)
        sys.exit(1)

    _log(f"Session name: {base_name}")

    # Current git branch
    current_branch = ""
    if has_git:
        rc, cb = _git_run(["rev-parse", "--abbrev-ref", "HEAD"], repo_root, capture=True)
        if rc == 0:
            current_branch = cb
    _log(f"Current git branch: {current_branch or 'none'}")

    # ── Idempotency check ────────────────────────────────────────────────────
    if sess_branch and sess_feat_dir:
        existing_dir = Path(repo_root) / sess_feat_dir if not Path(sess_feat_dir).is_absolute() else Path(sess_feat_dir)
        if current_branch == sess_branch and existing_dir.is_dir():
            _log(f"Already on branch '{sess_branch}' with feature dir '{sess_feat_dir}' — nothing to do")
            write_result(sess_branch, str(existing_dir), args.json)
            sys.exit(0)
        if current_branch and current_branch != sess_branch:
            _log(f"ERROR: Session expects branch '{sess_branch}' but current git branch is '{current_branch}'.")
            _log(f"       Switch to the correct branch  →  git checkout {sess_branch}")
            _log("       Or release the current feature first  →  /spec.release")
            sys.exit(1)

    # Fetch remote refs
    if has_git and not args.dry_run:
        subprocess.run(["git", "-C", repo_root, "fetch", "--all", "--prune"],
                       capture_output=True)

    # ── Resolve unique names ─────────────────────────────────────────────────
    branch_name = get_unique_branch_name(base_name, repo_root, has_git)
    folder_name = get_unique_folder_name(base_name, specs_dir)

    if branch_name != base_name:
        _log(f"Branch '{base_name}' taken — using '{branch_name}'")
    _log(f"Folder name: {folder_name}")

    feature_dir = specs_dir / folder_name
    spec_file   = feature_dir / "spec.md"

    # ── Create branch and spec dir ───────────────────────────────────────────
    if not args.dry_run:
        if has_git:
            _log(f"Creating git branch '{branch_name}'...")
            rc, _ = _git_run(["checkout", "-q", "-b", branch_name], repo_root)
            if rc != 0:
                rc2, _ = _git_run(["checkout", "-q", branch_name], repo_root)
                if rc2 != 0:
                    _log(f"ERROR: Failed to create or checkout '{branch_name}'")
                    sys.exit(1)
                _log(f"Switched to existing branch '{branch_name}'")
            else:
                _log(f"Branch '{branch_name}' created and checked out")
        else:
            _log("No git repo — skipping branch creation")

        feature_dir.mkdir(parents=True, exist_ok=True)

        if not spec_file.exists():
            tmpl = resolve_template("spec-template", repo_root)
            if tmpl and Path(tmpl).exists():
                import shutil
                shutil.copy2(tmpl, spec_file)
                _log(f"Spec file created from template: {spec_file}")
            else:
                spec_file.touch()
                _log(f"Spec file created (empty): {spec_file}")
        else:
            _log(f"Spec file already exists: {spec_file}")

        # Write branch_name and feature_dir back to session
        patch = json.dumps({
            "branch_name": branch_name,
            "feature_dir": f"specs/{folder_name}",
        })
        rc, _ = _run_manage(["--action", "update-multi", "--json-patch", patch])
        if rc != 0:
            sys.exit(rc)

        if args.agent_name:
            rc, _ = _run_manage(["--action", "add-agent", "--agent-name", args.agent_name])
            if rc != 0:
                sys.exit(rc)

        _log(f"Session updated: branch_name={branch_name}  feature_dir=specs/{folder_name}")
    else:
        _log(f"[dry-run] branch_name  → {branch_name}")
        _log(f"[dry-run] feature_dir  → specs/{folder_name}")
        _log("[dry-run] Would create session.json (if needed), branch and spec dir")

    write_result(branch_name, str(feature_dir), args.json)


if __name__ == "__main__":
    main()

File: .spec/scripts/create_new_feature.ts
#!/usr/bin/env node
import * as fs from "node:fs";
import * as path from "node:path";
import { spawnSync } from "node:child_process";
import { getRepoRoot, resolveTemplate, testHasGit } from "./common";

const scriptDir = __dirname;

/**
 * Writes verbose diagnostics to stderr for troubleshooting without changing stdout payloads.
 */
function logInfo(message: string, details?: unknown): void {
  if (details === undefined) {
    console.error(`[feature:debug] ${message}`);
    return;
  }
  console.error(`[feature:debug] ${message}`, details);
}

/**
 * Runs manage_session.ts and optionally captures stdout.
 */
function runManage(args: string[], capture = false): { code: number; out: string } {
  logInfo("Invoking manage_session", { args, capture });
  const commandArgs = ["tsx", path.join(scriptDir, "manage_session.ts"), ...args];
  const result = spawnSync("npx", commandArgs, {
    cwd: scriptDir,
    encoding: "utf-8",
    stdio: capture ? "pipe" : "inherit",
  });
  return { code: result.status ?? 1, out: (result.stdout ?? "").trim() };
}

/**
 * Emits user-facing flow logs.
 */
function log(msg: string): void {
  console.log(`[feature] ${msg}`);
}

/**
 * Executes a git command within the repository root.
 */
function gitRun(args: string[], repoRoot: string, capture = false): { code: number; out: string } {
  const gitResult = spawnSync("git", ["-C", repoRoot, ...args], {
    encoding: "utf-8",
    stdio: capture ? "pipe" : "ignore",
  });
  return { code: gitResult.status ?? 1, out: (gitResult.stdout ?? "").trim() };
}

function today(): string {
  const now = new Date();
  const pad2 = (value: number) => `${value}`.padStart(2, "0");
  return `${now.getFullYear()}${pad2(now.getMonth() + 1)}${pad2(now.getDate())}`;
}

/**
 * Finds an available date-prefixed branch name for the feature.
 */
function getUniqueBranchName(base: string, repoRoot: string, hasGit: boolean): string {
  // Always prefix branch with today's date: YYYYMMDD-<slug>
  const dateBase = `${today()}-${base}`;
  const takenBranchNames = new Set<string>();
  if (hasGit) {
    const localBranches = gitRun(["branch", "--list", `${dateBase}*`], repoRoot, true);
    if (localBranches.code === 0) {
      for (const line of localBranches.out.split("\n").filter(Boolean)) {
        takenBranchNames.add(line.trim().replace(/^\*/, "").trim());
      }
    }

    const remoteBranches = gitRun(["branch", "-r", "--list", `*/${dateBase}*`], repoRoot, true);
    if (remoteBranches.code === 0) {
      for (const line of remoteBranches.out.split("\n").filter(Boolean)) {
        const clean = line.trim();
        takenBranchNames.add(clean.split("/").at(-1) ?? clean);
      }
    }
  }

  // Use date-prefixed name; append counter suffix only on collision.
  if (!takenBranchNames.has(dateBase)) return dateBase;

  let counter = 2;
  while (true) {
    const candidate = `${dateBase}-${counter}`;
    if (!takenBranchNames.has(candidate)) return candidate;
    counter += 1;
  }
}

/**
 * Finds an available date-prefixed feature folder name.
 */
function getUniqueFolderName(base: string, specsDir: string): string {
  const prefix = `${today()}-${base}`;

  if (fs.existsSync(specsDir) && fs.statSync(specsDir).isDirectory()) {
    // Reuse an existing date-prefixed folder if it is empty (no files inside).
    const existing = path.join(specsDir, prefix);
    if (fs.existsSync(existing) && fs.statSync(existing).isDirectory()) {
      const hasFiles = fs.readdirSync(existing).length > 0;
      if (!hasFiles) return prefix; // reuse the empty folder
    }

    // Find the highest already-taken numbered variant to avoid collisions.
    const takenFolderNames = new Set<string>();
    for (const name of fs.readdirSync(specsDir)) {
      const full = path.join(specsDir, name);
      if (fs.statSync(full).isDirectory() && name.startsWith(prefix)) takenFolderNames.add(name);
    }

    if (!takenFolderNames.has(prefix)) return prefix;
    let counter = 2;
    while (true) {
      const candidate = `${prefix}-${counter}`;
      if (!takenFolderNames.has(candidate)) return candidate;
      counter += 1;
    }
  }

  return prefix;
}

/**
 * Prints result values in either JSON or human-friendly format.
 */
function writeResult(branch: string, featureDir: string, useJson: boolean): void {
  if (useJson) {
    console.log(JSON.stringify({ BRANCH_NAME: branch, FEATURE_DIR: featureDir, SPEC_FILE: `${featureDir}/spec.md` }));
  } else {
    console.log(`BRANCH_NAME: ${branch}`);
    console.log(`FEATURE_DIR: ${featureDir}`);
    console.log(`SPEC_FILE:   ${featureDir}/spec.md`);
  }
}

/**
 * Reads a single CLI argument value.
 */
function arg(name: string): string {
  const flagIndex = process.argv.indexOf(name);
  return flagIndex >= 0 && flagIndex + 1 < process.argv.length ? process.argv[flagIndex + 1] : "";
}

const name = arg("--name");
const description = arg("--description");
const agentName = arg("--agent-name");
const useJson = process.argv.includes("--json");
const dryRun = process.argv.includes("--dry-run");

logInfo("Starting feature creation flow", {
  name: name || null,
  descriptionProvided: Boolean(description),
  agentName: agentName || null,
  useJson,
  dryRun,
});

const repoRoot = getRepoRoot();
const hasGit = testHasGit(repoRoot);
const specsDir = path.join(repoRoot, ".spec", "specs");
logInfo("Resolved environment", { repoRoot, hasGit, specsDir });

if (!dryRun) fs.mkdirSync(specsDir, { recursive: true });

if (!dryRun) {
  const initArgs = ["--action", "init"];
  if (name) initArgs.push("--name", name);
  if (description) initArgs.push("--description", description);
  const init = runManage(initArgs);
  if (init.code !== 0) process.exit(init.code);
}

let sessionFeatureName = "";
let sessionBranch = "";
let sessionFeatureDir = "";

const multi = runManage(["--action", "get-multi", "--fields", "name,branch_name,feature_dir"], true);
if (multi.code === 0 && multi.out) {
  try {
    const sessionSnapshot = JSON.parse(multi.out) as { name?: string; branch_name?: string; feature_dir?: string };
    sessionFeatureName = sessionSnapshot.name ?? "";
    sessionBranch = sessionSnapshot.branch_name ?? "";
    sessionFeatureDir = sessionSnapshot.feature_dir ?? "";
  } catch {
    // Ignore malformed output.
  }
}

logInfo("Loaded current session snapshot", {
  sessionFeatureName: sessionFeatureName || null,
  sessionBranch: sessionBranch || null,
  sessionFeatureDir: sessionFeatureDir || null,
});

if (!sessionFeatureName) sessionFeatureName = name;
if (!sessionFeatureName) {
  const msg = dryRun
    ? "ERROR: No active session and --name not provided. Pass --name <slug> for dry-run."
    : "ERROR: Feature name is not set. Pass --name <slug> to provide one.";
  log(msg);
  process.exit(1);
}

log(`Session name: ${sessionFeatureName}`);

let currentBranch = "";
if (hasGit) {
  const cb = gitRun(["rev-parse", "--abbrev-ref", "HEAD"], repoRoot, true);
  if (cb.code === 0) currentBranch = cb.out;
}
log(`Current git branch: ${currentBranch || "none"}`);

// If session already points to the current branch + directory, this call is idempotent.
if (sessionBranch && sessionFeatureDir) {
  const existingDir = path.isAbsolute(sessionFeatureDir) ? sessionFeatureDir : path.join(repoRoot, sessionFeatureDir);
  if (currentBranch === sessionBranch && fs.existsSync(existingDir) && fs.statSync(existingDir).isDirectory()) {
    log(`Already on branch '${sessionBranch}' with feature dir '${sessionFeatureDir}' — nothing to do`);
    writeResult(sessionBranch, existingDir, useJson);
    process.exit(0);
  }
  if (currentBranch && currentBranch !== sessionBranch) {
    log(`ERROR: Session expects branch '${sessionBranch}' but current git branch is '${currentBranch}'.`);
    log(`       Switch to the correct branch  ->  git checkout ${sessionBranch}`);
    log("       Or release the current feature first  ->  /spec.release");
    process.exit(1);
  }
}

if (hasGit && !dryRun) {
  logInfo("Fetching remote branches for collision checks");
  spawnSync("git", ["-C", repoRoot, "fetch", "--all", "--prune"], { stdio: "ignore" });
}

const branchName = getUniqueBranchName(sessionFeatureName, repoRoot, hasGit);
const folderName = getUniqueFolderName(sessionFeatureName, specsDir);
logInfo("Resolved branch/folder targets", { branchName, folderName });
if (branchName !== sessionFeatureName) log(`Branch '${sessionFeatureName}' taken — using '${branchName}'`);
log(`Folder name: ${folderName}`);

const featureDir = path.join(specsDir, folderName);
const specFile = path.join(featureDir, "spec.md");

if (!dryRun) {
  if (hasGit) {
    log(`Creating git branch '${branchName}'...`);
    const createBranchResult = gitRun(["checkout", "-q", "-b", branchName], repoRoot);
    if (createBranchResult.code !== 0) {
      const switchBranchResult = gitRun(["checkout", "-q", branchName], repoRoot);
      if (switchBranchResult.code !== 0) {
        log(`ERROR: Failed to create or checkout '${branchName}'`);
        process.exit(1);
      }
      log(`Switched to existing branch '${branchName}'`);
    } else {
      log(`Branch '${branchName}' created and checked out`);
    }
  } else {
    log("No git repo — skipping branch creation");
  }

  fs.mkdirSync(featureDir, { recursive: true });
  logInfo("Ensured feature directory exists", { featureDir });

  if (!fs.existsSync(specFile)) {
    const tmpl = resolveTemplate("spec-template", repoRoot);
    if (tmpl && fs.existsSync(tmpl)) {
      fs.copyFileSync(tmpl, specFile);
      log(`Spec file created from template: ${specFile}`);
    } else {
      fs.closeSync(fs.openSync(specFile, "a"));
      log(`Spec file created (empty): ${specFile}`);
    }
  } else {
    log(`Spec file already exists: ${specFile}`);
  }

  const patch = JSON.stringify({ branch_name: branchName, feature_dir: `.spec/specs/${folderName}` });
  logInfo("Updating session with branch and feature dir", { patch });
  const upd = runManage(["--action", "update-multi", "--json-patch", patch]);
  if (upd.code !== 0) process.exit(upd.code);

  if (agentName) {
    const add = runManage(["--action", "add-agent", "--agent-name", agentName]);
    if (add.code !== 0) process.exit(add.code);
  }

  log(`Session updated: branch_name=${branchName}  feature_dir=.spec/specs/${folderName}`);
} else {
  log(`[dry-run] branch_name  -> ${branchName}`);
  log(`[dry-run] feature_dir  -> .spec/specs/${folderName}`);
  log("[dry-run] Would create session.json (if needed), branch and spec dir");
}

logInfo("Feature creation flow completed", { branchName, featureDir, useJson, dryRun });

writeResult(branchName, featureDir, useJson);

